### PR TITLE
[ENH/FIX] Re-adding database migrations

### PR DESCRIPTION
**Pull request by DESm1th**
I've scrapped all the old legacy migration scripts and gotten flask-migrate set up to successfully make a blank database with the right schema.

Still to fix:

1. I need to update our readme to include info on how to use and update the migrations

2. I've got to sort out the StudyUser 'site_only' issue. SQLAlchemy hates the partial index and has caused a rolling set of bugs that I've been playing whack-a-mole with and now I also can't get alembic/flask-migrate to properly translate it in the migration script. I'm thinking about abandoning it and just using a value 'all' to indicate access to all sites. This would mean either having a 'pseudo-site' in the sites table or removing the foreign key on the site_only column (meaning bad site values could creep in without proper input checking everywhere). I'd like to fix this issue before merging so the database created perfectly matches our production schema

### CODE REVIEW

**Comment by DESm1th** [comment]
Also, ignore that ^. the flake8 setting to ignore the migrations folder is bundled with this pull request

**Comment by DESm1th** [comment]
OK! This is ready for review now. As before, ignore all pep8 / codacy issues in migrations. The one pep8 'issue' in models is not a problem and will break the query if changed.

File: README.md [hunk 1]
```diff
@@ -21,33 +18,33 @@ QC metrics generated by our nightly pipelines.
...
-you are trying to get familiar with flask, with
+programmed in python using the [Flask](http://flask.pocoo.org/) framework (Strongly recommend
+[this tutorial](https://blog.miguelgrinberg.com/post/the-flask-mega-tutorial-part-i-hello-world) if
+you are trying to get familiar with flask) with
 [SQLAlchemy](https://www.sqlalchemy.org/). The web server itself is
 [NGINX](https://www.nginx.com/) with
 [uWSGI](https://uwsgi-docs.readthedocs.io/en/latest/). Authentication is handled
 with [OAuth](https://en.wikipedia.org/wiki/OAuth) and we currently support
-github and gitlab for login.
+github for login.
```
Comment (josephmje) [inline README.md]:
```suggestion
GitHub for login.
```

File: README.md [hunk 2]
```diff
@@ -171,28 +164,29 @@ In addition, you may need to create your user account yourself with
...
+createdb dashboard
+flask db upgrade
+```
 
-#### If you're in TIGRLab
-The easiest way to set up your test database is to just load in an old backup
-of the real database (ask @DESm1th for this). Once you have the backup
-you can set up your database by running the following as a user with database
-creation privileges
+If you're in TIGRLab and wish to work with a pre-populated database the easiest
```
Comment (josephmje) [inline README.md]:
```suggestion
If you're in TIGRLab and wish to work with a pre-populated database, the easiest
```

File: README.md [hunk 3]
```diff
@@ -171,28 +164,29 @@ In addition, you may need to create your user account yourself with
...
 # If no host is specified psql will connect to postgres on your machine
 psql dashboard -f $path_to_your_backup
 ```
 
-#### If you're not
-You'll have to build an empty database from scratch from the SQL files to be
-added later. Sorry - @DESm1th, May 24 2018
-
 #### Superuser Dashboard Access
 If this is a development set up you may also want to give your user account
```
Comment (josephmje) [inline README.md]:
```suggestion
If this is a development setup, you may also want to give your user account
```

File: README.md [hunk 4]
```diff
@@ -171,28 +164,29 @@ In addition, you may need to create your user account yourself with
...
-Superuser access to the database for convenience. To do this you must connect to
-the database as a user that already has superuser permissions (on most new
+superuser access to the database for convenience. To do this you must connect
```
Comment (josephmje) [inline README.md]:
```suggestion
superuser access to the database for convenience. To do this, you must connect
```

File: README.md [hunk 5]
```diff
@@ -171,28 +164,29 @@ In addition, you may need to create your user account yourself with
...
+to the database as a user that already has superuser permissions (on most new
 installations this would be the postgres user). Run the following in your
```
Comment (josephmje) [inline README.md]:
```suggestion
installations, this would be the postgres user). Run the following in your
```

File: README.md [hunk 6]
```diff
@@ -476,38 +470,38 @@ the backups and store them at `/mnt/backup/dashboard`, so we should have two
 copies at all times.
 
 ## Code Tips
-This is just a collection of small things to be aware of if you're going to write 
-code for the dashboard. To help with profiling and debugging consider using the 
+This is just a collection of small things to be aware of if you're going to write
+code for the dashboard. To help with profiling and debugging consider using the
```
Comment (josephmje) [inline README.md]:
```suggestion
code for the dashboard. To help with profiling and debugging, consider using the
```

Comment (josephmje) [reply README.md]: Maybe eventually, we can create a `CONTRIBUTING.md` file and move this content there.

File: README.md [hunk 7]
```diff
@@ -476,38 +470,38 @@ the backups and store them at `/mnt/backup/dashboard`, so we should have two
...
 chunks by saving html in another file and then importing it with something like
 ```jinja
 {% include 'my_other_file.html' %}
 ```
-performance wise this is not always a good idea. Each and every time the 'include' 
-statement is read when the website's page is loaded the included file has to be 
-read from the filesystem. File reads are (relatively) slow and if the include is 
-inside of a loop with a large number of iterations you can easily add extra seconds 
+performance wise this is not always a good idea. Each and every time the 'include'
+statement is read when the website's page is loaded the included file has to be
```
Comment (josephmje) [inline README.md]:
```suggestion
statement is read when the website's page is loaded, the included file has to be
```

File: README.md [hunk 8]
```diff
@@ -476,38 +470,38 @@ the backups and store them at `/mnt/backup/dashboard`, so we should have two
...
+read from the filesystem. File reads are (relatively) slow and if the include is
+inside of a loop with a large number of iterations you can easily add extra seconds
```
Comment (josephmje) [inline README.md]:
```suggestion
inside of a loop with a large number of iterations, you can easily add extra seconds
```

File: README.md [hunk 9]
```diff
@@ -476,38 +470,38 @@ the backups and store them at `/mnt/backup/dashboard`, so we should have two
...
 
 Some tips to get the most out of Jinja without adding too much overhead:
-  - If you have a loop and want to 'include' the body of the loop from another 
-    file, it's better to keep the loop inside the included file (so the file is 
+  - If you have a loop and want to 'include' the body of the loop from another
+    file, it's better to keep the loop inside the included file (so the file is
     opened and read once, rather than once per iteration)
   - If you have an 'if' statement and the body of it is included from another
-    snippet it's better to keep the 'if' in your original file, so you dont 
+    snippet it's better to keep the 'if' in your original file, so you dont
```
Comment (josephmje) [inline README.md]:
```suggestion
    snippet, it's better to keep the 'if' in your original file so you don't
```

File: README.md [hunk 10]
```diff
@@ -533,8 +527,18 @@ cmh = Site.query.get('CMH')
 for timepoint in cmh.timepoints:
    # Do some stuff with the timepoint record here
 ```
-This loop would generate one query to the database for each timepoint that has 
-the site 'CMH'. If code like that were embedded in a function called in a jinja 
-template then those extra queries would hit every time a user loaded the page. 
+This loop would generate one query to the database for each timepoint that has
+the site 'CMH'. If code like that were embedded in a function called in a jinja
+template then those extra queries would hit every time a user loaded the page.
```
Comment (josephmje) [inline README.md]:
```suggestion
template, then those extra queries would hit every time a user loaded the page.
```

File: migrations/alembic.ini [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+[handlers]
+keys = console
+
+[formatters]
+keys = generic
+
+[logger_root]
+level = WARN
+handlers = console
+qualname =
```
Comment (josephmje) [inline migrations/alembic.ini]: Do these unassigned variables need to exist?

Comment (DESm1th) [reply migrations/alembic.ini]: I believe so. It's an autogenerated file, I just changed the verbosity.

**Comment by josephmje** [comment]
Mostly added some small nitpicky comments to the README. Thanks for this and also all that documentation!

File: README.md (hunk 6 above)
Comment (DESm1th) [reply README.md]: Yeah definitely!

**Pull request merged**